[Draft] [serve] Prometheus metrics for AutoscalingConfig

arcyleung opened the pull request:
Implementation of AutoscalingConfig prometheus_custom_metrics: 
<img width="1335" height="1076" alt="image" src="https://github.com/user-attachments/assets/0df2b7e8-3bd0-406f-a75c-4dee31ad0f1a" />


## Why are these changes needed?

Support collection of Prometheus exporter metrics at each replica, and report them to the controller, if the serve config AutoscalingConfig specifies.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds optional per-replica Prometheus metric collection for autoscaling, refactors Prometheus helpers into a shared module, and updates proto/config/constants/replica logic and tests.
> 
> - **Serve autoscaling (backend)**:
>   - Enable collecting custom Prometheus metrics per replica when `AutoscalingConfig.prometheus_metrics` is set; fetched at `RAY_SERVE_REPLICA_AUTOSCALING_METRIC_PROMETHEUS_HOST` and merged into autoscaling reports.
>   - Extend replica metrics manager to asynchronously fetch/filter metrics by `replica` label and include alongside ongoing-requests/custom metrics.
> - **Config/Proto**:
>   - Add `prometheus_metrics` to `AutoscalingConfig` (Python) and new proto messages `PrometheusCustomMetrics`/`PrometheusMetric`; map list<str> ↔ proto on (de)serialization.
>   - Update JSON surfaces and defaults accordingly.
> - **Constants**:
>   - Introduce `RAY_SERVE_REPLICA_AUTOSCALING_METRIC_PROMETHEUS_HOST` env var for exporter host.
> - **Prometheus utilities**:
>   - New `python/ray/_private/prometheus_utils.py` with fetch/parse/filter helpers and async fetch; remove duplicated logic from `test_utils` and update imports across tests/benchmarks.
> - **Tests/BUILD**:
>   - Add `test_prometheus_autoscaling_metrics.py`; adjust existing tests and BUILD to include env and new imports; minor test stability tweaks.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 02a540ce4c277c15ece9b3f2a21d46270f1ebfdd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

abrarsheikh commented on python/ray/serve/_private/replica.py:
```diff
@@ -376,6 +398,95 @@ def _push_autoscaling_metrics(self) -> Dict[str, Any]:
+                                        metric_samples_by_name[sample.name] = []
+                                    metric_samples_by_name[sample.name].append(sample)
+
+                            # Extract requested metrics
+                            for metric_name, promql_expression in prometheus_metrics:
+                                if metric_name in metric_samples_by_name:
+                                    # Get the latest sample for this metric
+                                    samples = metric_samples_by_name[metric_name]
+                                    if samples:
+                                        # Use the last sample's value
+                                        latest_sample = samples[-1]
+                                        metrics_result[metric_name] = float(
+                                            latest_sample.value
+                                        )
+                                    else:
+                                        metrics_result[metric_name] = 0.0
+                                else:
+                                    logger.warning(
+                                        f"Metric {metric_name} not found in exporter response"
+                                    )
+                                    metrics_result[metric_name] = 0.0
```
+1, also i dont think promql_expression is used. Also I dont think we would be able to leverage it given we are getting metrics from local exporter ?

jjyao commented on python/ray/_private/prometheus_utils.py:
```diff
@@ -0,0 +1,266 @@
+"""Utilities for fetching and parsing Prometheus metrics.
```
Assuming this is used by libraries? If so lets move it to `_common` folder.

harshit-anyscale commented on python/ray/_common/prometheus_utils.py:
```diff
@@ -0,0 +1,266 @@
+
+async def fetch_prometheus_metrics_async(
+    prom_address: str,
+    timeout: Optional[float] = None,
```
we should put a default timeout here

harshit-anyscale commented on python/ray/_common/prometheus_utils.py:
```diff
@@ -0,0 +1,266 @@
+    Returns:
+        Raw metrics text or None if fetch failed.
+    """
+    import aiohttp
```
lets move it to top of file

harshit-anyscale commented on python/ray/serve/_private/replica.py:
```diff
@@ -376,6 +391,54 @@ def _push_autoscaling_metrics(self) -> Dict[str, Any]:
+            prom_addr = RAY_SERVE_REPLICA_AUTOSCALING_METRIC_PROMETHEUS_HOST
+            logger.debug(f"Fetching metrics from prometheus exporter at {prom_addr}")
+
+            metrics_text = await fetch_prometheus_metrics_async(
```
is there any prom API which can return us the exact metrics we are looking for? IIUC, this will fetch all prom metrics in our memory, and then we will find the metrics which are helpful to us.

arcyleung commented:
## Update (agent pass)

Pushed tip `b28a362` on `arcyleung:prometheus-metrics` with:

- **Node-local Prometheus query cache/dedup** (in-process + on-disk TTL + file lock) so co-located replicas do not N× hit the same host/query
- Host scheme normalization (fixes double `http://`); empty default host with explicit error when `prometheus_metrics` set without env
- Controller probe uses `RAY_SERVE_REPLICA_AUTOSCALING_METRIC_PROMETHEUS_HOST` (no hardcoded `localhost:9090`)
- Unified `set_autoscaling_config` so missing Prom host does not skip other metrics collection
- Tests assert **metric values**, not only key presence; added `test_prometheus_query_cache.py`
- Local multi-node docker harness under `python/ray/serve/tests/multi_node_prometheus/`

**Note on full rebase onto latest master:** a clean replay onto current `ray-project/ray` master exists locally (`prometheus-metrics-rebased`, `git merge-base --is-ancestor upstream/master HEAD` passes) but force-pushing that history is blocked by the PAT lacking the `workflow` scope (new upstream file `.github/workflows/notify-slack-on-label.yml`). Re-push of the full rebased history needs a token with `workflow` (or SSH key on the account).

Cursor/Gemini comment themes addressed on tip are documented in the agent review notes (DI handler, double URL prefix, hardcoded host, value extraction, early-return metrics, etc.).

Commit:
[serve] 2/2 Prometheus autoscaling metrics tests

Core unit coverage: value extraction, co-located cache dedup, Prom error
status handling, zero-TTL, and replica fetch path with co-located ids.

Signed-off-by: Arthur Leung <arcyleung@gmail.com>

Commit:
[serve] 1/2 Prometheus autoscaling metrics

Add AutoscalingConfig.prometheus_metrics and replica-side PromQL collection
with a node-local TTL cache so co-located replicas sharing host/query do not
N× hit Prometheus. Configured expressions are queried as-is. Prometheus JSON
status=error raises instead of being treated as empty samples.

Signed-off-by: Arthur Leung <arcyleung@gmail.com>

arcyleung commented:
## Review disposition (rebased tip)

Human review threads (abrarsheikh / jjyao / harshit) are addressed on the current tip or declined with rationale:

- **Done:** `prometheus_metrics` Field+description; PromQL utils in `ray._common.prometheus_utils`; unified `set_autoscaling_config` (start/stop pusher); empty host errors; default 10s timeout; co-located node-local cache; queries as-is (no forced `replica=` rewrite).
- **Declined rename** to `prometheus_custom_metrics` — keep shorter public name; list values are PromQL/names (see field description).
- **Histogram/counters:** use PromQL in the list (e.g. `histogram_quantile(...)`); we extract instant vector/scalar samples as floats.

Also on tip: raise on Prometheus JSON `status=error` (do not cache as empty); removed unused Serve-side TTL constant (env still honored in `_common`).

Rebased onto latest `ray-project/ray` master; tip will be the SHA on this branch after push.
